dcos/dcos-ui-update-service
# DCOS-46390 Abstraction to watch zk nodes

The main focus of this PR was to introduce the `ValueNodeWatcher` & `ParentNodeWatcher` abstractions to the `zookeeper` package. These will be used when resolving the update race-condition between masters.

To dog food at least one of these I replaced the polling logic that existed in `zkVersionStore` with a `ValueNodeWatcher`. I also introduced a `ParentNodeWatcher` during testing but removed before commiting this work.

The two watchers are similar but since they have different data (`[]byte` & `[]string`) and slightly different behavior they are separate implementations.

There are some changes to the `zookeeper` `client` to accommodate the watchers. But also the change to get the version before `Set` & `Delete` calls was made after reading and better understanding the `zookeeper` API docs.

## Testing

Tests
1. Run unit tests `go test -race ./...`

Running locally
1. Run the service locally `docker-compose up`
- You should see lots of `trace` logs related to the version watcher
2. Update the zookeeper node to a new value (2.24.4 & 2.37.0 are version that exist in default universe). I do this using a [node.js script](https://gist.github.com/TattdCodeMonkey/880c0826e060da007275ad532cf94d94). When running with docker the zookeeper server is available @ `localhost:2181` the version node is `/dcos/ui-update/version`
- You should see the change detected in the logs and an update triggered if the value is different. If you update to a package that exists the update should be successfully completed.

## Trade-offs



## Dependencies

N/A

## Screenshots

N/A
